File: coco/pattern/springsec/hsm-bulk-service/src/main/java/com/hsm/bulk/crypto/IvFactory.java
package com.hsm.bulk.crypto;

import org.bouncycastle.crypto.CryptoServicesRegistrar;

import java.security.SecureRandom;

/** Duplicated verbatim from com.hsm.core.crypto.IvFactory -- see DekManager's class comment. */
public final class IvFactory {

    public static final int IV_LENGTH_BYTES = 12; // 96-bit recommended for AES-GCM

    private static final SecureRandom RANDOM = CryptoServicesRegistrar.getSecureRandom();

    private IvFactory() {
    }

    public static byte[] generate() {
        byte[] iv = new byte[IV_LENGTH_BYTES];
        RANDOM.nextBytes(iv);
        return iv;
    }
}

File: coco/pattern/springsec/hsm-databricks-udf/src/hsm_databricks_udf/iv_factory.py
"""
IV generation — 96-bit (12 bytes) random IV per NIST SP 800-38D §8.2.
Each call to generate() produces a fresh cryptographically secure value.

Vendored from app/crypto/iv_factory.py — see DATABRICKS_UDF_DESIGN.md §5 for
why this is copied rather than imported: keeps this package's dependency
footprint to just `cryptography` + stdlib, decoupled from app/'s full
FastAPI/SQLAlchemy tree. Keep byte-for-byte identical to the source if that
file changes.
"""

import secrets


IV_LENGTH_BYTES = 12  # 96-bit recommended for AES-GCM


def generate() -> bytes:
    return secrets.token_bytes(IV_LENGTH_BYTES)
